packages/@lwc/ssr-runtime/src/index.ts

```diff
@@ -0,0 +1,161 @@
+        const set = new Set(className.split(MULTI_SPACE).filter(Boolean));
+        for (const newClassName of newClassNames) {
+            set.add(newClassName);
+        }
+        this.el.className = Array.from(set).join(' ');
+    }
+
+    contains(className: string) {
+        const currentClassNameStr = this.el.className;
+        return currentClassNameStr.split(MULTI_SPACE).includes(className);
+    }
+
+    remove(...classNamesToRemove: string[]) {
+        const className = this.el.className;
+        const set = new Set(className.split(MULTI_SPACE).filter(Boolean));
+        for (const newClassName of classNamesToRemove) {
+            set.delete(newClassName);
+        }
+        this.el.className = Array.from(set).join(' ');
+    }
+
+    replace(oldClassName: string, newClassName: string) {
+        let classWasReplaced = false;
+        const className = this.el.className;
+        const listOfClasses = className.split(MULTI_SPACE).filter(Boolean);
+        listOfClasses.forEach((value, idx) => {
+            if (value === oldClassName) {
+                classWasReplaced = true;
+                listOfClasses[idx] = newClassName;
+            }
+        });
+        this.el.className = listOfClasses.join(' ');
+        return classWasReplaced;
+    }
+
+    toggle(classNameToToggle: string, force?: boolean) {
+        const classNameStr = this.el.className;
+        const set = new Set(classNameStr.split(MULTI_SPACE).filter(Boolean));
+        if (!set.has(classNameToToggle) && force !== false) {
+            set.add(classNameToToggle);
+        } else if (set.has(classNameToToggle) && force !== true) {
+            set.delete(classNameToToggle);
+        }
+        this.el.className = Array.from(set).join(' ');
+        return set.has(classNameToToggle);
+    }
+}
+
+export class LightningElement {
+    isConnected = false;
+    className = '';
+    // TODO [W-14977927]: protect internals from userland
+    __attrs?: Record<string, string>;
+    __classList: ClassList | null = null;
+
+    constructor(propsAvailableAtConstruction: Record<string, any>) {
+        Object.assign(this, propsAvailableAtConstruction);
+    }
+
+    // TODO [W-14977927]: protect internals from userland
+    __internal__setState(
+        props: Record<string, any>,
+        reflectedProps: string[],
+        attrs: Record<string, any>
+    ) {
+        Object.assign(this, props);
+        this.__attrs = attrs;
+
+        // Whenever a reflected prop changes, we'll update the original props object
+        // that was passed in. That'll be referenced when the attrs are rendered later.
+        for (const reflectedPropName of reflectedProps) {
+            Object.defineProperty(this, reflectedPropName, {
+                get() {
+                    return props[reflectedPropName];
+                },
+                set(newValue) {
+                    props[reflectedPropName] = newValue;
+                },
+                enumerable: true,
```

Comment:
Why set enumerable here and not for other properties?

Reply:
Some background: the compiler currently looks at which prop names are being used in the component JS, then only includes compiled output relevant for the component code. This might not catch 100% of cases that are used in client-side LWC. However, I think that's a reasonable trade-off, since we've already articulated that SSR will involve some limitations that do not exist for client-side LWC.

As for why I made this enumerable... I don't actually know. It doesn't hurt anything, but there isn't necessarily a reason to do it either. I guess it seems less surprising for a property to be enumerable that is known to be accessed in a `LightningElement`.

Do you have any strong opinions on this? I think this is just a thing I did, and didn't think about too much.